src/coreclr/jit/morphblock.cpp

```diff
@@ -0,0 +1,1551 @@
+                        {
+                            // We can't clone multiple copies of a tree with persistent side effects
+                            noway_assert((m_dstAddr->gtFlags & GTF_PERSISTENT_SIDE_EFFECTS) == 0);
+
+                            dstAddrClone = m_comp->gtCloneExpr(m_dstAddr);
+                            noway_assert(dstAddrClone != nullptr);
+
+                            JITDUMP("dstAddr - Multiple Fields Clone created:\n");
+                            DISPTREE(dstAddrClone);
+
+                            // Morph the newly created tree
+                            dstAddrClone = m_comp->fgMorphTree(dstAddrClone);
+                        }
+
+                        // Is the address of a local?
+                        GenTreeLclVarCommon* lclVarTree = nullptr;
+                        bool                 isEntire   = false;
+                        bool*                pIsEntire  = (m_blockSizeIsConst ? &isEntire : nullptr);
+                        if (dstAddrClone->DefinesLocalAddr(m_comp, m_blockSize, &lclVarTree, pIsEntire))
+                        {
+                            lclVarTree->gtFlags |= GTF_VAR_DEF;
+                            if (!isEntire)
+                            {
+                                lclVarTree->gtFlags |= GTF_VAR_USEASG;
+                            }
+                        }
+                    }
+                }
+
+                LclVarDsc* srcVarDsc      = m_comp->lvaGetDesc(m_srcLclNum);
+                unsigned   srcFieldLclNum = srcVarDsc->lvFieldLclStart + i;
+                LclVarDsc* srcFieldVarDsc = m_comp->lvaGetDesc(srcFieldLclNum);
+
+                // Have to set the field sequence -- which means we need the field handle.
+                CORINFO_CLASS_HANDLE classHnd = srcVarDsc->GetStructHnd();
+                CORINFO_FIELD_HANDLE fieldHnd =
+                    m_comp->info.compCompHnd->getFieldInClass(classHnd, srcFieldVarDsc->lvFldOrdinal);
+                FieldSeqNode* curFieldSeq = m_comp->GetFieldSeqStore()->CreateSingleton(fieldHnd);
+
+                unsigned  srcFieldOffset = m_comp->lvaGetDesc(srcFieldLclNum)->lvFldOffset;
+                var_types srcType        = srcFieldVarDsc->TypeGet();
+
+                if (!m_dstUseLclFld)
+                {
+
+                    if (srcFieldOffset == 0)
+                    {
+                        m_comp->fgAddFieldSeqForZeroOffset(dstAddrClone, curFieldSeq);
+                    }
+                    else
+                    {
+                        GenTree* fieldOffsetNode = m_comp->gtNewIconNode(srcFieldVarDsc->lvFldOffset, curFieldSeq);
+                        dstAddrClone = m_comp->gtNewOperNode(GT_ADD, TYP_BYREF, dstAddrClone, fieldOffsetNode);
+                    }
+
+                    dstFld = m_comp->gtNewIndir(srcType, dstAddrClone);
+                }
+                else
+                {
+                    assert(m_dstAddOff == 0);
```

Comment:
there was a potential bug with trees like `ADD(CNST X, ADD CNST Y, ADDR(LCL_VAR Z))` when we were morphing an assignment to field by field using the local var. In the past, we were ignoring `X+Y` offset for it. Lucky for us we don't have such cases.